### ISSUE TITLE
Not possible to change screen when app inactive

### ISSUE THREAD

**Issue by rmtmckenzie** [flutter/flutter#30745]
This is somewhere between a 'how do I' and a feature request. What I'm trying to do is essentially block app's screen whenever it isn't the active app (i.e. switcher, when the user pulls down the notification pane on iOS, etc). I've set up a WidgetsBindingObserver to detect when the app switches to Inactive or Paused, and I'm trying to render a different page in that case.

I'm fairly sure this isn't possible within flutter as I've tried using `SchedulerBinding.instance`'s `scheduleForcedFrame` and `scheduleWarmupFrame` which don't work, and they seem like the most reasonable thing to use. With either, the frame isn't actually rendered until the app resumes again.

I don't know if this is actually even possible to do - but it does seem like it is possible with native iOS code. I think `[self.view snapshotViewAfterScreenUpdates:YES];` and possibly `setNeedsDisplay`. I haven't looked into android too much yet.

If anyone has an idea of how this can be done completely within flutter I'd appreciate it, otherwise I'll be looking at doing it natively...

**Comment by MaterialTime** [comment]
I think you want to use have a StatefulWidget with WidgetBindingObserver and override didChangeAppLifecycleState

**Comment by rmtmckenzie** [comment]
Maybe I should have been clearer - that's what I did.  

> I've set up a WidgetsBindingObserver

When I made a change after the app goes inactive, it didn't actually render until after the app resumed. I was trying to make it render before the app actually goes into the background (i.e. the app switcher)

**Comment by escamoteur** [comment]
Hi, sorry for the delay, I just want to check if you could solve this problem.

**Comment by rmtmckenzie** [comment]
@escamoteur well on iOS I ended up writing a bit of code that essentially blurs the UIWindow natively, and I haven't managed to get around to doing the same on Android yet. But if it were to be considered as something to do make possible within flutter that would be great as the blurring on iOS isn't the greatest user experience as it's not really tied to the flutter lifecycle and I really don't know what to do on Android yet.

**Comment by escamoteur** [comment]
Do you want want to change the page out of security reasons?

**Comment by rmtmckenzie** [comment]
Yes. The result of being able to force a frame to be run when the app is inactive would essentially allow for control of what is shown on the "snapshot" screen that is shown in the app switcher, so that I could either block out the entire screen or more ideally just the particular relevant items that need to be blanked out for security reasons.

**Comment by gaaclarke** [comment]
Couldn't you just listen to the lifecycle events and swap out the view goes between `inactive` and `resumed`?

Observer of lifecycle events example: https://github.com/flutter/flutter/blob/master/examples/layers/services/lifecycle.dart

That's what the documentation I saw for UIKit apps seems to suggest.

**Comment by rmtmckenzie** [comment]
The issue is more that I wanted to be able to detect that the screen was about to go inactive and then render a different screen (blurred out) for the render directly from flutter. The flutter engine at the time (and I think still currently) defers rendering as soon as it gets the inactive event, and I couldn't even force trigger a render. I haven't checked this recently though.

**Comment by rmtmckenzie** [comment]
That or even allow for a forced render using scheduleForcedFrame or something else like it after the inactive event - I can see why rendering is stopped when the window becomes inactive for performance/battery reasons - this is a bit of a corner-case, albeit a nice to have one.

**Comment by pickalex** [comment]
hello ,did you solve this problem in flutter ?

**Comment by shabanov-krl** [comment]
i think i have same issue

usually everything works correctly, but sometimes redrawing does not occur on ios (tested on ios 17 only in release build mode)

all prints works as expected

```dart
import 'dart:ui';

import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: ScreenBlurOverlay(
        child: MyHomePage(),
      ),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key});

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> with WidgetsBindingObserver {
  @override
  void initState() {
    super.initState();

    WidgetsBinding.instance.addObserver(this);
  }

  @override
  void dispose() {
    WidgetsBinding.instance.removeObserver(this);

    super.dispose();
  }

  @override
  void didChangeAppLifecycleState(AppLifecycleState state) {
    if (state == AppLifecycleState.resumed) {
      ScreenBlurOverlay.of(context).disable();
    } else {
      ScreenBlurOverlay.of(context).enable();
    }
  }

  @override
  Widget build(BuildContext context) {
    return const Scaffold(
      body: Center(
        child: Text('text'),
      ),
    );
  }
}

abstract interface class BlurController {
  bool get enabled;

  void enable();
  void disable();
}

class ScreenBlurOverlay extends StatefulWidget {
  final Widget child;

  const ScreenBlurOverlay({
    required this.child,
    super.key,
  });

  static BlurController of(BuildContext context) =>
      context.getInheritedWidgetOfExactType<_InheritedScreenBlurOverlay>()!.controller;

  @override
  State<ScreenBlurOverlay> createState() => _ScreenBlurOverlayState();
}

class _ScreenBlurOverlayState extends State<ScreenBlurOverlay> implements BlurController {
  bool _enabled = false;

  @override
  void enable() {
    if (_enabled) {
      print('already enabled');
      return;
    }

    setState(() {
      _enabled = true;
    });
    print('enabled');
  }

  @override
  void disable() {
    if (!_enabled) {
      print('already disabled');
      return;
    }

    setState(() {
      _enabled = false;
    });
    print('disabled');
  }

  @override
  bool get enabled => _enabled;

  @override
  Widget build(BuildContext context) {
    print('build $_enabled');
    return _InheritedScreenBlurOverlay(
      controller: this,
      child: Stack(
        children: [
          widget.child,
          (() {
            print('child $_enabled');
            return const SizedBox();
          })(),
          Center(child: Text(_enabled.toString())),
          if (_enabled) ...[
            (() {
              print('filter $_enabled');
              return const SizedBox();
            })(),
            BackdropFilter(
              filter: ImageFilter.blur(sigmaX: 5, sigmaY: 5),
              child: const SizedBox(
                height: double.infinity,
                width: double.infinity,
              ),
            ),
          ],
        ],
      ),
    );
  }
}

class _InheritedScreenBlurOverlay extends InheritedWidget {
  final BlurController controller;

  const _InheritedScreenBlurOverlay({
    required this.controller,
    required super.child,
  });

  @override
  bool updateShouldNotify(_InheritedScreenBlurOverlay oldWidget) => false;
}
```

https://github.com/flutter/flutter/assets/46450095/de95ae82-5cda-439e-98b1-6c3bef0ce52e

https://github.com/flutter/flutter/assets/46450095/99072220-5534-4c96-9320-4394016fd78f

**Comment by AslamThachapalli** [comment]
@shabanov-krl The solution works!
Abstracted and created a package around it.
Hope it may help someone.

Package: [shield_view](https://pub.dev/packages/shield_view)

**Comment by shabanov-krl** [comment]
> @shabanov-krl The solution works! Abstracted and created a package around it. Hope it may help someone.
> 
> Package: [shield_view](https://pub.dev/packages/shield_view)

@AslamThachapalli what about those floating bugs that I described above?

**Comment by buchen** [comment]
@shabanov-krl @AslamThachapalli 

To add: The code above and the shield_view package do **not** work for me. Yes, it blurs, if I move into the task switcher by swiping up from the bottom. But it does not blur when the user for example switches the phone off. If the application goes directly into the "paused" state, there is no paint event after the WidgetsBindingObserver received the event - hence no blur.

I found two packages that use native code to achieve this: [screen_protector](https://pub.dev/packages/screen_protector) and [secure_application](https://pub.dev/packages/secure_application). The former works for me, but I am unsure how to unlock with biometrics. The latter does work and supports unlock, but does not allow to disable the lock again (only parts of my app need the protection).
